sqlsmith: specify random index hints

yuzefovich opened the pull request:
**sql: add is_inverted column to crdb_internal.table_indexes vtable**

Release note (sql change): A new boolean column 'is_inverted' has been
added to crdb_internal.table_indexes virtual table which indicates
whether the index is inverted or not.

**sqlsmith: specify random index hints**

Now in 50% a random index on a table will be chosen to be used as the
hint. Inverted indexes are omitted.

Release note: None

yuzefovich committed:
sqlsmith: specify random index hints

Now in 50% a random index on a table will be chosen to be used as the
hint. Inverted indexes are omitted.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/43102)
<!-- Reviewable:end -->

mjibson commented:
This is failing because it is using index hints during INSERT.

yuzefovich commented:
Ok, the build is now green.

mjibson commented on pkg/internal/sqlsmith/schema.go:
```diff
@@ -57,13 +62,31 @@ func (s *Smither) ReloadSchemas() error {
+	if s.coin() {
+		indexNames := make([]tree.Name, 0, len(indexes))
+		for _, index := range indexes {
+			if !index.Inverted && !strings.Contains(index.Name.String(), "jsonb") {
```
do you need the second condition here? you can't make a non-inverted jsonb index, only inverted.

mjibson commented on pkg/internal/sqlsmith/relational.go:
```diff
@@ -41,22 +41,28 @@ func makeSchemaTable(s *Smither, refs colRefs, forJoin bool) (tree.TableExpr, co
 }
 
-func (s *Smither) getSchemaTable() (tree.TableExpr, *tree.TableName, *tableRef, colRefs, bool) {
+func (s *Smither) getSchemaTable(
```
I would prefer to have two functions that do this instead of having one that is passed a boolean argument. One should be able to call the other and either add or remove the index hint. This should cleanup the call site changes.

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43102)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson)

___
*[pkg/internal/sqlsmith/relational.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43102#-LvloGqcL-g1aEH_6tmM-r1-48:-LvnJknf06BZ5AinB0sF:bl1h72v) ([raw file](https://github.com/cockroachdb/cockroach/blob/f575a402d8ba85c53634554acbdad418283014b5/pkg/internal/sqlsmith/relational.go#L48)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I would prefer to have two functions that do this instead of having one that is passed a boolean argument. One should be able to call the other and either add or remove the index hint. This should cleanup the call site changes.
</blockquote></details>

Ok, done.
___
*[pkg/internal/sqlsmith/schema.go, line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43102#-LvloGqcL-g1aEH_6tmN-r1-77:-LvnKxeI-ZgxghME8r9n:b3ncimh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f575a402d8ba85c53634554acbdad418283014b5/pkg/internal/sqlsmith/schema.go#L77)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

do you need the second condition here? you can't make a non-inverted jsonb index, only inverted.
</blockquote></details>

Unfortunately `index.Inverted` field is not populated, so the first condition here is actually useless. I briefly looked into the schema introspection and didn't find where I could get that information from, so the simplest (albeit hacky) way is to check explicitly for "jsonb" in the name of the index. Without the second check, the rate of errors like `pq: index "seed__jsonb_idx" is inverted and cannot be used for this query` is quite high.


<!-- Sent from Reviewable.io -->

mjibson commented:
Index definitions are populated at https://github.com/cockroachdb/cockroach/blob/b2fcc8cb865f5f7c49ab9158e9d5570c471873b5/pkg/internal/sqlsmith/schema.go#L205. Can you teach them about inverted indexes? That should fix this problem.

yuzefovich reviewed (commented):
I looked into that, and `IndexDescriptor.Type` (which describes whether the index is forward or inverted) doesn't seem to be exposed through the SQL interface in any way (I looked through all reads of that field, and all of them seem to be internal). We could extend `information_schema.statistics` to include that field which would allow us to add another "Type" column to the output of `SHOW INDEXES`, but I'm not sure if it'll be breaking some PG-compatibility. Probably not, right?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43102#-:-Lvs3PBx83dU0sxaK7Fb:brbchm1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson)



<!-- Sent from Reviewable.io -->

mjibson commented:
@jordanlewis do you know if there's a way to introspect and get the inverted flag on an index?

jordanlewis commented:
No, it seems like we missed adding that to any of the catalog tables. I think the way this is supposed to work is that `pg_am` (access method) is supposed to contain one entry per type of index, roughly, and indexes are supposed to refer to a particular access method.

The easiest way to fix this would be to add it to `crdb_internal.table_indexes`. It can get a new column for `is_inverted`.

yuzefovich committed:
sql: add is_inverted column to crdb_internal.table_indexes vtable

Release note (sql change): A new boolean column 'is_inverted' has been
added to crdb_internal.table_indexes virtual table which indicates
whether the index is inverted or not.

mjibson commented on pkg/internal/sqlsmith/schema.go:
```diff
@@ -248,6 +272,35 @@ func extractIndexes(
 					Direction: dir,
 				})
 			}
+			row, err := db.Query(fmt.Sprintf(`
```
if you expect exactly one row you can use `db.QueryRow('query').Scan(&blah)`

yuzefovich replied:
Oh, nice, thanks!

yuzefovich commented:
I think the build should be green now, PTAL.

yuzefovich commented:
TFTR!

bors r+
